src/main/java/uk/gov/pay/ledger/transaction/service/TransactionService.java

```diff
@@ -88,4 +101,50 @@ public void upsertTransactionFor(EventDigest eventDigest) {
         TransactionEntity transaction = transactionEntityFactory.create(eventDigest);
         transactionDao.upsert(transaction);
     }
+
+    public TransactionEventResponse findTransactionEvents(String externalId, String gatewayAccountId) {
+        Map<String, TransactionEntity> transactionEntityMap = getTransactionsAsMap(externalId, gatewayAccountId);
+
+        if (transactionEntityMap.isEmpty()) {
+            throw new BadRequestException(format("Transaction with id [%s] not found", externalId));
+        }
+
+        List<TransactionEvent> transactionEvents = getTransactionEventsFor(transactionEntityMap);
+
+        return TransactionEventResponse.of(externalId, removeDuplicates(transactionEvents));
+    }
+
+    private Map<String, TransactionEntity> getTransactionsAsMap(String externalId, String gatewayAccountId) {
+        return transactionDao.findTransactionByExternalOrParentIdAndGatewayAccountId(
+                externalId, gatewayAccountId)
+                .stream()
+                .collect(Collectors.toMap(TransactionEntity::getExternalId,
+                        transactionEntity -> transactionEntity));
+    }
+
+    private List<TransactionEvent> getTransactionEventsFor(Map<String, TransactionEntity> transactionEntityMap) {
+        List<Event> events = eventDao.findEventsForExternalIds(transactionEntityMap.keySet());
+        return mapToTransactionEvent(transactionEntityMap, events);
+    }
+
+    private List<TransactionEvent> removeDuplicates(List<TransactionEvent> transactionEvents) {
+        // removes 1. events without mapping to transaction state
+        // 2. duplicate events based on external_id,resource_type & state (gets the first created event based on event date)
+        Collection<Optional<TransactionEvent>> values = transactionEvents.stream()
+                .filter(transactionEvent -> transactionEvent.getState() != null)
+                .collect(groupingBy(transactionEvent -> transactionEvent.getExternalId() + transactionEvent.getResourceType() + transactionEvent.getState(),
+                        Collectors.minBy(Comparator.comparing(TransactionEvent::getTimestamp))))
+                .values();
+
+        return values.stream().filter(Optional::isPresent)
+                .map(Optional::get)
+                .sorted(Comparator.comparing(TransactionEvent::getTimestamp))
+                .collect(Collectors.toList());
+    }
+
+    private List<TransactionEvent> mapToTransactionEvent(Map<String, TransactionEntity> transactionEntityMap, List<Event> eventList) {
```

Comment:
[minor] from reading code point of view this method should be above `removeDuplicates`.

Reply:
👍 Updated